src/Chainweb/Pact/PactService.hs

```diff
@@ -477,97 +478,134 @@ finalizeCheckpointer finalize = do
 _liftCPErr :: Either String a -> PactServiceM cas a
 _liftCPErr = either internalError' return
 
-type Balances = HM.HashMap Text Decimal
 
 data Uniqueness = Duplicate | Unique
 
+-- | Performs a dry run of PactExecution's `buyGas` function for transactions being validated.
+--
+attemptBuyGas
+    :: PayloadCas cas
+    => Checkpointer
+    -> Miner
+    -> Maybe BlockHash
+    -> Vector (ChainwebTransaction, Uniqueness)
+    -> PactServiceM cas (Vector (ChainwebTransaction, Uniqueness, Bool))
+attemptBuyGas cp miner nonGenesisParentHash txs = withDiscardedBatch $ do
+  mbLatestBlock <- liftIO $ _cpGetLatestBlock cp
+  (bh, bhash) <- case mbLatestBlock of
+                     Nothing -> throwM NoBlockValidatedYet
+                     (Just !p) -> return p
+  let target = Just (succ bh, bhash)
+  withCheckpointer target "validateChainwebTxs/attemptBuyGas" $ \(PactDbEnv' dbEnv) -> do
+      (T2 _ mcache) <- runCoinbase nonGenesisParentHash dbEnv miner
+      psEnv <- ask
+      res <- mapM (runBuyGas psEnv dbEnv mcache) txs
```

Comment:
We need to fold the mcache through the various txs here, as loading capabilities could cause other modules to be loaded. See `applyPactCmds` for how this is done.

Reply:
So store the ModuleCache that's returned from `runGas` (https://github.com/kadena-io/chainweb-node/blob/master/src/Chainweb/Pact/TransactionExec.hs#L476) in each transaction and pass it to the other transactions?

Reply:
See latest commit